File: next-greatest-element-II.py
# time - O(3n)
# space - O(n)
# des- similar to the problem of daily temperature, but have to traverse the array again(2n).
# trick- using mod to get the index of the elements in the nums array
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        if not nums:
            return []
        stack = []
        res = [-1] * len(nums)
        for i in range(2 * len(nums)):
            while stack and nums[i % len(nums)] > nums[stack[len(stack) - 1]]:
                res[stack.pop()] = nums[i % len(nums)]
            if i < len(nums):
                stack.append(i)
        return res

File: NextGreatestElementII.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {

        //O(n) space and O(3n) exact time
        int[] result = new int[nums.length];

        if (nums == null) return result;

        Arrays.fill(result, -1);

        Stack<Integer> stack = new Stack<>();
        int n = nums.length;

        for (int i=0; i< 2 *n; i++) {

            while (!stack.isEmpty() && nums[i % n] > nums[stack.peek()]) {
                int top = stack.pop();
                result[top] = nums[i%n];
            }
            if (i < n) {
                stack.push(i);
            }

        }

        return result;

    }
}